Add tests for proposed WebDriver Shadow DOM support

jimevans opened the pull request:
The tests within this PR correspond to the changes proposed by the [PR to the WebDriver Specification for supporting Shadow DOM](https://github.com/w3c/webdriver/pull/1565)

jimevanssfdc committed:
Add tests for proposed WebDriver Shadow DOM support

jimevanssfdc committed:
Correcting HTTP verb for WebDriver command

jimevanssfdc committed:
Adding WebDriver transport commands for Shadow DOM

whimboo commented on webdriver/tests/find_element_from_shadow_root/find.py:
```diff
@@ -0,0 +1,47 @@
+from tests.support.asserts import assert_success
+
+
+def get_shadow_root(session, element):
```
The tests in this file are testing `Find Element`. So for cases where we have to retrieve the shadow root, the appropriate webdriver client implementation (`element.shadow_root`) should be used.

jimevans replied:
Done in 94b8624

whimboo commented on webdriver/tests/find_element_from_shadow_root/find.py:
```diff
@@ -0,0 +1,47 @@
+
+
+@pytest.fixture
+def get_checkbox_dom(inline):
```
Please move this fixture into a `conftest.py` file in that directory. For `performActions` we named the fixture for the test page `test_actions_page`. Maybe we can do it similar here.

We could also add this HTML file under `support/html/shadow_dom.html`, so that we can share it between all the different shadow DOM related APIs.

jimevans replied:
Done in 94b8624

whimboo commented on webdriver/tests/find_element_from_shadow_root/find.py:
```diff
@@ -0,0 +1,47 @@
+def test_find_element(session, get_checkbox_dom):
+    session.url = get_checkbox_dom
+    custom_element = session.find.css("custom-checkbox-element", all=False)
+    shadow_root = get_shadow_root(session, custom_element).value
```
As mentioned above this can just be `custom_element.shadow_root`.

jimevans replied:
Done in 94b8624

whimboo commented on webdriver/tests/find_element_from_shadow_root/find.py:
```diff
@@ -0,0 +1,47 @@
+    custom_element = session.find.css("custom-checkbox-element", all=False)
+    shadow_root = get_shadow_root(session, custom_element).value
+    response = find_element(session, shadow_root, "css", "input")
+    assert_success(response)
```
Make sure to check that we got the correct element returned, eg. it should be a checkbox.

Also could you duplicate most of the tests from `find_element/find.py`? Would be great to also have tests for missing parameters, non-existing browsing contexts and such. That should be too much work to get the existing tests adapted. Also `user_prompts.py` would be important.

jimevans replied:
Added additional tests, including user prompt tests, in subsequent commits.

whimboo commented on webdriver/tests/find_elements_from_shadow_root/find.py:
```diff
@@ -0,0 +1,47 @@
+    custom_element = session.find.css("custom-checkbox-element", all=False)
+    shadow_root = get_shadow_root(session, custom_element).value
+    response = find_elements(session, shadow_root, "css", "input")
+    assert_success(response)
```
Same comments as above for `Find Element` apply for this test file.

jimevans replied:
Done in 694cc74

whimboo commented on webdriver/tests/get_element_shadow_root/get.py:
```diff
@@ -0,0 +1,45 @@
+    session.url = inline("<div><p>no shadow root</p></div>")
+    element = session.find.css("div", all=False)
+    response = get_shadow_root(session, element)
+    assert_error(response, "no such shadow root")
```
Same comments as above for `Find Element` apply for this test file.

jimevans replied:
Done in 694cc74

whimboo reviewed (changes requested):
Thanks @jimevans for taking a stab on it. I have a couple of comments that I added inline. I would appreciate to have good test coverage for scenarios we are already aware of for other commands. I hope that this will be fine with you.

jimevanssfdc committed:
Updating WebDriver client with code review suggestions

jimevans commented on tools/webdriver/webdriver/client.py:
```diff
@@ -372,6 +372,40 @@ def from_json(cls, json, session):
+        body = {"using": strategy,
+                "value": selector}
+        url = "shadow/%s/element" % (self.id)
+        return self.session.send_session_command("POST", url, body)
```
Done in c4e6a2c

jimevans commented on tools/webdriver/webdriver/client.py:
```diff
@@ -804,6 +838,10 @@ def selected(self):
     def screenshot(self):
         return self.send_element_command("GET", "screenshot")
 
+    @command
```
Done in c4e6a2c

jimevans commented on tools/webdriver/webdriver/client.py:
```diff
@@ -372,6 +372,40 @@ def from_json(cls, json, session):
+class ShadowRoot(object):
+    identifier = "shadow-075b-4da1-b6ba-e579c2d3230a"
+
+    def __init__(self, id, session):
```
Done in c4e6a2c

jimevanssfdc committed:
Fleshing out find element from shadow root tests per review

jimevanssfdc committed:
Renaming test_shadow_page to get_shadow_page

Doing this to avoid potential naming convention of `test_` prefix in pytest.

Also fix linting error

jimevanssfdc committed:
Fleshing out tests for find elements from shadow root

jimevanssfdc committed:
Add user prompt tests for find element(s) from shadow root

jimevanssfdc committed:
Update Get Element Shadow Root tests

Includes adding user prompt tests

AutomatedTester commented on webdriver/tests/find_element_from_shadow_root/user_prompts.py:
```diff
@@ -0,0 +1,129 @@
+
+
+@pytest.fixture
+def check_user_prompt_not_closed_but_exception(session, create_dialog, get_shadow_page):
```
Please move the fixtures to the `conftest.py` file

jimevans replied:
Done in 9e70d8f.

whimboo replied:
Note that we don't do that for all the other commands. Also these fixtures are just for the user prompt tests, and don't give any value for any other test in this folder.

jimevans replied:
@whimboo @AutomatedTester I’m happy to put them wherever anyone wants them, but let’s come to a consensus on them.

jimevans replied:
On second thought, I'm moving them back to the test files for the sake of consistency. There are also issues with the fixtures referencing stuff that isn't imported or existing in the `conftest.py`. If we want to make a more global change to affect all user prompt test files, we can move them all in one set of changes.

AutomatedTester commented on tools/webdriver/webdriver/client.py:
```diff
@@ -372,6 +372,42 @@ def from_json(cls, json, session):
+        return cls(uuid, session)
+
+    def send_shadow_command(self, method, uri, body=None):
+        url = "shadow/%s/%s" % (self.id, uri)
```
Please change this to 
```suggestion
        url = "shadow/{}/{}".format(self.id, uri)
```

jimevans replied:
Done in 5720604

jimevans replied:
or rather, botched in 5720604 and fixed in 5d3b4c5

AutomatedTester commented on webdriver/tests/find_element_from_shadow_root/user_prompts.py:
```diff
@@ -0,0 +1,129 @@
+
+
+@pytest.fixture
+def check_user_prompt_closed_with_exception(session, create_dialog, get_shadow_page):
```
Please move the fixtures to the conftest.py file

jimevans replied:
Done in 9e70d8f.

AutomatedTester commented on webdriver/tests/find_element_from_shadow_root/user_prompts.py:
```diff
@@ -0,0 +1,129 @@
+
+
+@pytest.fixture
+def check_user_prompt_closed_without_exception(session, create_dialog, get_shadow_page):
```
Please move the fixtures to the conftest.py file

jimevans replied:
Done in 9e70d8f.

AutomatedTester commented on webdriver/tests/find_elements_from_shadow_root/user_prompts.py:
```diff
@@ -0,0 +1,131 @@
+
+
+@pytest.fixture
+def check_user_prompt_closed_with_exception(session, create_dialog, get_shadow_page):
```
Please move the fixtures to the conftest.py file

jimevans replied:
Done in 9e70d8f.

AutomatedTester commented on webdriver/tests/find_elements_from_shadow_root/user_prompts.py:
```diff
@@ -0,0 +1,131 @@
+
+
+@pytest.fixture
+def check_user_prompt_not_closed_but_exception(session, create_dialog, get_shadow_page):
```
Please move the fixtures to the conftest.py file

jimevans replied:
Done in 9e70d8f.

AutomatedTester commented on webdriver/tests/get_element_shadow_root/user_prompts.py:
```diff
@@ -0,0 +1,118 @@
+
+
+@pytest.fixture
+def check_user_prompt_closed_without_exception(session, create_dialog, get_checkbox_dom):
```
Please move the fixtures to the conftest.py file

jimevans replied:
Done in 9e70d8f.

AutomatedTester commented on webdriver/tests/get_element_shadow_root/user_prompts.py:
```diff
@@ -0,0 +1,118 @@
+
+
+@pytest.fixture
+def check_user_prompt_closed_with_exception(session, create_dialog, get_checkbox_dom):
```
Please move the fixtures to the conftest.py file

jimevans replied:
Done in 9e70d8f.

whimboo replied:
Not sure why my comment from late yesterday hasn't made it here, but I'm against this move given that none of the other test files in this folder benefit from those fixtures, and specifically it would diverge from the way how all the other user prompt tests have been implemented. If we decide to move fixtures out, we should do it all together.

AutomatedTester commented on webdriver/tests/get_element_shadow_root/user_prompts.py:
```diff
@@ -0,0 +1,118 @@
+
+
+@pytest.fixture
+def check_user_prompt_not_closed_but_exception(session, create_dialog, get_checkbox_dom):
```
Please move the fixtures to the conftest.py file

jimevans replied:
Done in 9e70d8f.

jimevanssfdc committed:
Code review: moving fixtures to conftest.py files

jimevanssfdc committed:
Code review: Updating string format mechanism

jimevanssfdc committed:
Correct string formatting for real

whimboo commented on webdriver/tests/find_element_from_shadow_root/conftest.py:
```diff
@@ -0,0 +1,71 @@
+@pytest.fixture
+def get_shadow_page(inline, shadow_content):
+    return inline("""
```
I actually wonder how this can work with the current implementation. What you return here is a string as produced by `inline()`. In the tests you call `get_shadow_page(...)`, which has to fail on a string.

So what you want is to return an inner function; the same as what we do for most of the global fixtures. And only the inner method has the `shadow_content` argument.

This also applies to the other commands.

jimevans replied:
Acknowledged. Will work on this.

jimevans replied:
Took a stab at this in c9921ba. I've probably botched it, so do feel free to re-review.

whimboo commented on webdriver/tests/find_element_from_shadow_root/find.py:
```diff
@@ -0,0 +1,122 @@
+        {"using": using, "value": value})
+
+
+def test_found_element_equivalence(session, get_shadow_page):
```
Mind moving this test down right before `test_find_element` so that we start with the null value test as we do for all the other tests?

jimevans replied:
Done for find element and find elements in d74e42a.

whimboo commented on webdriver/tests/find_element_from_shadow_root/find.py:
```diff
@@ -0,0 +1,122 @@
+    shadow_root = custom_element.shadow_root
+
+    response = find_element(session, shadow_root.id, using, value)
+    assert_error(response, "no such element")
```
I assume we also need a test when `shadow_root` is stale.

jimevans replied:
Added in 1cf10f0

whimboo commented on webdriver/tests/get_element_shadow_root/get.py:
```diff
@@ -0,0 +1,54 @@
+    session.url = get_checkbox_dom
+    custom_element = session.find.css("custom-checkbox-element", all=False)
+    response = get_shadow_root(session, custom_element)
+    assert_success(response)
```
Please add an assertion that checks for the correct shadow root.

jimevans replied:
Assertion added in 8ef520e. Slightly convoluted, in that we compare the `host` of the shadow root to the element we're requesting the shadow root for, but since we don't have a way to pass a shadow root back via JS to compare, that's the best approach I can think of. Feel free to make another suggestion.

whimboo replied:
Hm, shouldn't we able to serialize a shadow root element as retrieved by `Execute Script` similar to a [window or frame object](https://searchfox.org/mozilla-central/source/testing/web-platform/tests/webdriver/tests/execute_script/json_serialize_windowproxy.py)?

jimevans replied:
Maybe, but that's not how it's specified in the spec language. Plus a Shadow Root "element" isn't an element, it's a different thing. We can, in theory, serialize the window and frame proxies, but no implementation [actually does](https://wpt.fyi/results/webdriver/tests/execute_script/json_serialize_windowproxy.py?label=master&label=experimental&aligned). That would also require changes in the `Execute Script` spec language to support it, which wasn't part of this effort. If, in the future, we decide to make this serializable via `Execute Script`, we can update these tests to utilize that mechanism.

whimboo replied:
So I mixed that up. Sorry. In that case lets resolve it.

whimboo reviewed (changes requested):
It's all great work. Thanks Jim! I have some more comments here. Most apply to all of the different commands.

jimevanssfdc committed:
Rearrange tests per code review

jimevanssfdc committed:
Code review: Have get_shadow_page return internal function

jimevanssfdc committed:
Revert move of user prompt fixtures to conftest.py

jimevanssfdc committed:
Adding forgotten file from previous commit

jimevanssfdc committed:
Add tests for detached shadow root

jimevanssfdc committed:
Add assertion for returning correct shadow root

jimevanssfdc committed:
Add import statement to conftest.py files

jimevanssfdc committed:
Reverting rename of inner function

jimevanssfdc committed:
Correcting URL variable parsing in get element shadow root

jimevanssfdc committed:
Add new exceptions to WebDriver error parsing

whimboo commented on webdriver/tests/get_element_shadow_root/conftest.py:
```diff
@@ -0,0 +1,22 @@
+import pytest
+
+@pytest.fixture
+def get_checkbox_dom(inline):
```
Maybe remove the `get_` prefix here because it's not returning a method.

jimevans replied:
Done in eda55bd

jimevanssfdc committed:
Adding same element assert for find element from shadow root

jimevanssfdc committed:
Code review: changing name of supporing fixture

jimevans commented on webdriver/tests/find_element_from_shadow_root/find.py:
```diff
@@ -0,0 +1,132 @@
+    shadow_root = custom_element.shadow_root
+    response = find_element(session, shadow_root.id, using, value)
+    assert_success(response)
+
```
Added in fb8ab3c

whimboo reviewed (approved):
This looks all fine to me now (by not being able to run any of these tests). Thanks a lot.
